larq_zoo/resnet_e.py

```diff
@@ -0,0 +1,182 @@
+    downsample = x.get_shape().as_list()[-1] != filters
+
+    if downsample:
+        residual = keras.layers.AvgPool2D(pool_size=2, strides=2)(x)
+        residual = keras.layers.Conv2D(
+            filters, kernel_size=1, use_bias=False, kernel_initializer="glorot_normal"
+        )(residual)
+        residual = keras.layers.BatchNormalization(momentum=0.9, epsilon=1e-5)(residual)
+    else:
+        residual = x
+
+    x = lq.layers.QuantConv2D(
+        filters,
+        kernel_size=3,
+        strides=strides,
+        padding="same",
+        input_quantizer=args.quantizer,
+        kernel_quantizer=args.quantizer,
+        kernel_constraint=args.constraint,
+        kernel_initializer="glorot_normal",
+        use_bias=False,
+        metrics=[],
+    )(x)
+    x = keras.layers.BatchNormalization(momentum=0.9, epsilon=1e-5)(x)
+
+    return keras.layers.add([x, residual])
+
+
+@registry.register_model
+def resnet_e(args, dataset, input_tensor=None, include_top=True):
+    input = utils.get_input_layer(dataset.input_shape, input_tensor)
+
+    if dataset.input_shape[0] and dataset.input_shape[0] < 50:
+        x = keras.layers.Conv2D(
+            args.initial_filters,
+            kernel_size=3,
+            padding="same",
+            kernel_initializer="he_normal",
+            use_bias=False,
+        )(input)
+    else:
+        x = keras.layers.Conv2D(
+            args.initial_filters,
+            kernel_size=7,
+            strides=2,
+            padding="same",
+            kernel_initializer="he_normal",
+            use_bias=False,
+        )(input)
+
+        x = keras.layers.BatchNormalization(momentum=0.9, epsilon=1e-5)(x)
+        x = keras.layers.Activation("relu")(x)
+        x = keras.layers.MaxPool2D(3, strides=2, padding="same")(x)
+        x = keras.layers.BatchNormalization(momentum=0.9, epsilon=1e-5)(x)
+
+    for block, (layers, filters) in enumerate(zip(*args.spec)):
+        # this tricks adds shortcut connections between original resnet blocks
+        # we multiple number of blocks by 2, but add only one layer instead of two in each block
+        for layer in range(layers * 2):
+            strides = 1 if block == 0 or layer != 0 else 2
+            x = residual_block(x, args, filters, strides=strides)
+
+    if include_top:
+        x = keras.layers.Activation("relu")(x)
+        x = keras.layers.GlobalAvgPool2D()(x)
+        x = keras.layers.Dense(
+            dataset.num_classes,
+            activation="softmax",
+            kernel_initializer="glorot_normal",
+        )(x)
+    return keras.Model(inputs=input, outputs=x)
+
+
+@registry.register_hparams(resnet_e)
+class default(HParams):
+    epochs = 120
+    batch_size = 1024
+    num_layers = 18
+    learning_rate = 0.004
+    learning_factor = 0.3
+    learning_steps = [70, 90, 110]
+    initial_filters = 64
+    quantizer = lq.quantizers.SteSign(clip_value=1.25)
+    constraint = lq.constraints.WeightClip(clip_value=1.25)
+
```

Comment:
Why is the clip_value 1.25? Is this something from the supplementary material?

Reply:
That's what they used in the paper and the [supplementary materials](https://owncloud.hpi.de/s/1jrAUnqRAfg0TXH) also explores different clipping values.